// @aws-amplify/platform-core/lib/package_json_reader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.packageJsonSchema = exports.PackageJsonReader = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const zod_1 = __importDefault(require("zod"));
const errors_1 = require("./errors");
/**
 * return the package json
 */
class PackageJsonReader {
    read = (absolutePackageJsonPath) => {
        if (!fs_1.default.existsSync(absolutePackageJsonPath)) {
            throw new Error(`Could not find a package.json file at ${absolutePackageJsonPath}`);
        }
        let jsonParsedValue;
        try {
            jsonParsedValue = JSON.parse(
            // we have to use sync fs methods here because this is also used during cdk synth
            fs_1.default.readFileSync(absolutePackageJsonPath, 'utf-8'));
        }
        catch (err) {
            throw new errors_1.AmplifyUserError('InvalidPackageJsonError', {
                message: `Could not parse the contents of ${absolutePackageJsonPath}`,
                resolution: `Ensure that ${absolutePackageJsonPath} exists and is a valid JSON file`,
            }, err);
        }
        return exports.packageJsonSchema.parse(jsonParsedValue);
    };
    /**
     * Returns the contents of the package.json file in process.cwd()
     *
     * If no package.json file exists, or the content does not pass validation, an error is thrown
     */
    readFromCwd = () => {
        return this.read(path_1.default.resolve(process.cwd(), 'package.json'));
    };
}
exports.PackageJsonReader = PackageJsonReader;
/**
 * Type for package.json content.
 *
 * Add additional validation if there are other fields we need to read
 */
exports.packageJsonSchema = zod_1.default.object({
    name: zod_1.default.string().optional(),
    version: zod_1.default.string().optional(),
    type: zod_1.default.union([zod_1.default.literal('module'), zod_1.default.literal('commonjs')]).optional(),
});
