// @aws-amplify/backend-output-schemas/lib/auth/v1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.authOutputSchema = void 0;
const zod_1 = require("zod");
// The values stored in the stack output is converted to string.
exports.authOutputSchema = zod_1.z.object({
    version: zod_1.z.literal('1'),
    payload: zod_1.z.object({
        authRegion: zod_1.z.string(),
        userPoolId: zod_1.z.string(),
        webClientId: zod_1.z.string(),
        identityPoolId: zod_1.z.string(),
        amazonClientId: zod_1.z.optional(zod_1.z.string()),
        appleClientId: zod_1.z.optional(zod_1.z.string()),
        facebookClientId: zod_1.z.optional(zod_1.z.string()),
        googleClientId: zod_1.z.optional(zod_1.z.string()),
        allowUnauthenticatedIdentities: zod_1.z.optional(zod_1.z.string()),
        usernameAttributes: zod_1.z.string().optional(),
        signupAttributes: zod_1.z.string().optional(),
        passwordPolicyMinLength: zod_1.z.string().optional(),
        passwordPolicyRequirements: zod_1.z.string().optional(),
        mfaConfiguration: zod_1.z.string().optional(),
        mfaTypes: zod_1.z.string().optional(),
        verificationMechanisms: zod_1.z.string().optional(),
        socialProviders: zod_1.z.string().optional(),
        oauthDomain: zod_1.z.string().optional(),
        oauthScope: zod_1.z.string().optional(),
        oauthRedirectSignIn: zod_1.z.string().optional(),
        oauthRedirectSignOut: zod_1.z.string().optional(),
        oauthClientId: zod_1.z.string().optional(),
        oauthResponseType: zod_1.z.string().optional(),
    }),
});
